### PR TITLE
NIFI-7745: Add SampleRecord processor

### PR DESCRIPTION
**Pull request by mattyb149**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Adds a SampleRecord processor to select records from an incoming flow file based on a sampling strategy. The set of strategies in this PR are:

Interval Sampling (every Nth record)
Probabilistic Sampling (each record has a probability P of being chosen)
Reservoir Sampling (A sample of size K with each record having equal probability of being chosen)

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by ottobackwards** [comment]
This looks really cool.  I think an additional details documentation piece that illustrates an example  of what you would get from a given set of settings, for a given input and why would be a big help for users

**Comment by mattyb149** [comment]
Good point, I added the doc, please let me know if you'd like to see more/different details than what I have provided :)

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SampleRecord.java [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+
+    static final AllowableValue INTERVAL_SAMPLING = new AllowableValue(INTERVAL_SAMPLING_KEY, "Interval Sampling",
+            "Selects every Nth record where N is the value of the 'Interval Value' property");
+    static final AllowableValue PROBABILISTIC_SAMPLING = new AllowableValue(PROBABILISTIC_SAMPLING_KEY, "Probabilistic Sampling",
+            "Selects each record with probability P where P is the value of the 'Selection Probability' property");
+    static final AllowableValue RESERVOIR_SAMPLING = new AllowableValue(RESERVOIR_SAMPLING_KEY, "Reservoir Sampling",
+            "Creates a sample of K records where each record has equal probability of being included, where K is "
+                    + "the value of the 'Reservoir Size' property");
+
+    static final Validator INTEGER_PERCENT_VALIDATOR = (subject, value, context) -> {
```
Comment (jfrazee) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SampleRecord.java:82]: Could you just use the `StandardValidators.createLongValidator()`?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SampleRecord.java [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+            .name("sample-record-interval")
+            .displayName("Sampling Interval")
+            .description("Specifies the number of records to skip before writing a record to the outgoing FlowFile. This property is only "
+                    + "used if Sampling Strategy is set to Interval Sampling. A value of zero (0) will cause all records to be included in the"
+                    + "outgoing FlowFile.")
+            .required(false)
+            .addValidator(StandardValidators.NON_NEGATIVE_INTEGER_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .build();
+    static final PropertyDescriptor SAMPLING_PROBABILITY = new PropertyDescriptor.Builder()
```
Comment (jfrazee) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SampleRecord.java:134]: Would it make sense to add another property setting the random seed?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SampleRecord.java (hunk 1 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SampleRecord.java:82]: Good catch, I was looking for a range validator and missed it! Will update

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SampleRecord.java (hunk 2 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SampleRecord.java:134]: Yes (most likely for testing though), will add the property and documentation

**Comment by ottobackwards** [comment]
+1 fwiw, good stuff

**Comment by jfrazee** [comment]
@mattyb149 I went through this pretty closely and I think it looks good. The Probabilistic and Reservoir sampling produces the right distributions over 100k files. Two things I wanted to verify: (1) since the reservoir sampling is in-memory (for good reasons), setting it too high will be problematic? and (2) you've opted to not do these as controller services because that's potentially more complicated than most users would need/want?

**Comment by mattyb149** [comment]
You're right on both counts.  Setting a very large reservoir can cause memory issues but I chose to do it in-memory for performance reasons, figuring that sampling is meant to choose a (potentially much) smaller set of records, and if the total number of sampled records is still very large, the flowfile could be split beforehand and then sampled. It changes the overall behavior of course but I figured most of the use cases are covered by this implementations.
Also I didn't make the sampling strategies controller services (or even separate class files) mostly because I didn't foresee any other usage of them, but they can always be refactored into their own files and/or components if need be.

**Comment by ottobackwards** [comment]
I think the processor should have the resource attribute and there be warnings in the addition documentation / property descriptions then?

**Comment by mattyb149** [comment]
Good point, will add

**Pull request closed without merging**